db/dohast.c

```diff
@@ -0,0 +1,568 @@
+/*
+   Copyright 2015 Bloomberg Finance L.P.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+
+#include "comdb2.h"
+#include "sqliteInt.h"
+#include "vdbeInt.h"
+#include "ast.h"
+#include "dohsql.h"
+#include "sql.h"
+
+#define ast_verbose bdb_attr_get(thedb->bdb_attr, BDB_ATTR_DOHAST_VERBOSE)
+
+static void node_free(dohsql_node_t **pnode, sqlite3 *db);
+
+char *generate_columns(sqlite3 *db, ExprList *c, const char **tbl)
+{
+    char *cols = NULL;
+    Expr *expr = NULL;
+    int i;
+
+    for (i = 0; i < c->nExpr; i++) {
+        expr = c->a[i].pExpr;
+        if (expr->op != TK_COLUMN) {
+            if (cols)
+                sqlite3DbFree(db, cols);
+            return NULL;
+        }
+        if (!cols)
+            cols = sqlite3_mprintf("%s.\"%s\"%s%s%s", expr->pTab->zName,
+                                   c->a[i].pExpr->u.zToken,
+                                   (c->a[i].zName) ? " AS \"" : "",
+                                   (c->a[i].zName) ? c->a[i].zName : "",
+                                   (c->a[i].zName) ? "\" " : "");
+        else
+            cols = sqlite3_mprintf("%s, %s.\"%s\"%s%s%s", cols,
+                                   expr->pTab->zName, c->a[i].pExpr->u.zToken,
+                                   (c->a[i].zName) ? " AS \"" : "",
+                                   (c->a[i].zName) ? c->a[i].zName : "",
+                                   (c->a[i].zName) ? "\" " : "");
+        if (!cols)
+            return NULL;
+        if (tbl)
+            *tbl = expr->pTab->zName;
+    }
+
+    return cols;
+}
+
+static char *describeExprList(Vdbe *v, const ExprList *lst)
+{
+    char *ret;
+    char *tmp;
+    char *newterm;
+    int i;
+
+    ret = sqlite3ExprDescribe(v, lst->a[0].pExpr);
```

Comment:
Handle `sortOrder` or return `NULL` to disable parallel exec for `order by ... desc` if this is a `TODO`